Heuristic and LSTM model added 20200126

henryhao1991 opened the pull request:
Added the models and results for heuristic and LSTM as baselines. 

dataloader.py: Including classes to load data and feed to the model. But I wrote 2 functions, one for train/val/test split dataloader, one is just a single dataloader, since the heuristic model doesn't need to train and thus not necessary to split the dataset. I feel there might be a better way to do this than writing two separate functions.

model.py: Including heuristic model and LSTM model.

heuristic_baseline.py: Just a class to run the heuristic model and get results. Maybe I will change or add more functions later for the convenience of comparing results with other methods. Probably I need to do the same for the LSTM model, instead of write the whole training code in Jupyter notebooks...

Two Jupyter notebooks: Results from two models. I'm not sure if this is a good way to show results. Also, I'm still investigating LSTM model, and experimenting different setup (e.g., batch_size, different metrics, etc.) now.

Also, I tend to have a bad habit of not adding many comments in the codes. I will add more comments to the existing code (and start doing so for the rest of the project!), so that they will be easier to read.

henryhao1991 committed:
Readme changed

henryhao1991 committed:
dataloader added

henryhao1991 committed:
model added: heuristic and lstm

henryhao1991 committed:
heuristic results and bug fixes

henryhao1991 committed:
minor bug fixes

henryhao1991 committed:
LSTM finished building and bug fixes

henryhao1991 committed:
Added LSTM training results

TensorDuck commented on dataloader.py:
```diff
@@ -0,0 +1,103 @@
+
+
+class PadSequence:
+    def __call__(self, batch):
```
Question: Why is this a class and not a function?

henryhao1991 replied:
Good point. I was looking for a solution to pad sequence and train LSTM in batch, and this was found on one of the discussion in PyTorch forum. Now when I think about it, this was pass to collate_fn in dataloader. Maybe it is better to use function instead.

TensorDuck commented on dataloader.py:
```diff
@@ -0,0 +1,103 @@
+
+        labels = [x[1] for x in sorted_batch]
+        labels_padded = torch.nn.utils.rnn.pad_sequence(labels, batch_first=True)
+        return sequences_padded, labels_padded
```
The most important thing about working in an engineering team is making sure the code is transparent enough to be clear. I would recommend adding a docstring for this function and comments that clarify the purpose of each step with regards to the larger project, not just the features that are being manipulated.

henryhao1991 replied:
Yes, definitely I will add doc string and more comments to clarify codes. I think I can finish adding doc string and comments by Tuesday.

TensorDuck commented on dataloader.py:
```diff
@@ -0,0 +1,103 @@
+
+def split_dataloader(batch_size=1, total_num_games=-1, p_val=0.1, p_test=0.2, seed=3154, shuffle=True):
+    
+    dataset = PreprocessedParsedReplayDataset()
```
This class has keyword args that you aren't using. If they are allowed, they should be passed in in `split_datdaloader` as well.

TensorDuck commented on dataloader.py:
```diff
@@ -0,0 +1,103 @@
+
+
+def split_dataloader(batch_size=1, total_num_games=-1, p_val=0.1, p_test=0.2, seed=3154, shuffle=True):
+    
```
Add a docstring here.

TensorDuck commented on dataloader.py:
```diff
@@ -0,0 +1,103 @@
+    
+    train_loader = DataLoader(dataset, batch_size=batch_size, sampler=sample_train, collate_fn=PadSequence())
+    val_loader = DataLoader(dataset, batch_size=batch_size, sampler=sample_val, collate_fn=PadSequence())
+    test_loader = DataLoader(dataset, batch_size=batch_size, sampler=sample_test, collate_fn=PadSequence())
```
You're basically trying to make a train test split here right? Scikit-learn has a function that handles this:
https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html

Which you can use twice to make a train/test/validation split:
https://datascience.stackexchange.com/questions/15135/train-test-validation-set-splitting-in-sklearn

It is always better to use something pre-built then trying to write it yourself. It'll save the amount of mental overhead required to read code if there's just a clear 2-lines of function call rather than 10 lines of custom code doing the same thing.

TensorDuck commented on dataloader.py:
```diff
@@ -0,0 +1,103 @@
+        features = np.loadtxt(feature_path)
+        labels = np.loadtxt(label_path)
+        
+        return torch.Tensor(features), torch.Tensor(labels)
```
This is a very minor and subtle point: This is returning features and labels in a specific order. If someone were to use the dataloader, and then gets back an object, they would HAVE to know the order of features and labels.

I.e. You could use a dictionary instead:
```
return {"features": torch.Tensor(features), "labels": torch.Tensor(labels)}
```
This way, all someone needs to know is that there will be features and labels, not the order. This is particularly challenging to read since starting at the notebook file where it's used (where I start reading), I have to go 3 functions down to see what's being returned. You can think of the dictionary acting like a mini-class to organize the results with keywords.

TensorDuck commented on heuristic_baseline.py:
```diff
@@ -0,0 +1,44 @@
+    """
+    The class to run the heuristic model. No training needed.
+    Parameters can be tuned in model.heuristic
+    """
```
This seems like a well written class, but it seems overly specific to the heuristic model. My instinct is this can be generalized to not just the heuristic model, but any other model. I.E., it outputs metrics like accuracy and graphs.

TensorDuck commented on heuristic_baseline.py:
```diff
@@ -0,0 +1,44 @@
+        self.num_data = len(self.dataloader)
+        self.model = heuristic()
+        
+    def get_inputs_and_results(self, features, labels):
```
If this were a more generalized method, features/labels should be stored inside `self.model`.

henryhao1991 replied:
This function is more or less a function to preprocess data for heuristic model, and features/labels are from dataloader, which I feel is usually not stored inside the model. However, I think this can be written as another collate_fn for dataloader, just like the PadSequence for LSTM. But I'm not exactly sure and need to test it first.

TensorDuck replied:
I agree with you. On closer look, I think I misinterpreted what features and labels meant. In my head, I thought they were model features, not input features to make predictions on. 

In that case, this should be fine as is. Let's plan to talk more about what code needs to be generalized at our check-in this week.

henryhao1991 replied:
Yeah, that was a bad name for the function now I think about it. 

I'm also planning to discuss with you in detail on the generalization of the code. Right now, some generalized functions for each model in my head: 1. train(): train the model on the training set. Which should do nothing for the heuristic model, since it doesn't need to train. 2. get_accuracy():
get the accuracy on the test set, probably at different time instead of just at the end (which is what it is right now) 3. get_single_graph(): give a parsed game file as input, output the predicted probability at different times. This will be used as the deliverable, as the user may want to check the win prob graph for a specific game.

Sorry for the lengthy comment, but this is what's in my mind so far. We can discuss if these make sense or not during the check-in.

TensorDuck replied:
Definitely, let's talk more during the check-in. Seems like you're on the right track :thumbsup:, so let's break down the design a little more live.

TensorDuck commented on heuristic_baseline.py:
```diff
@@ -0,0 +1,44 @@
+    def __init__(self):
+        self.dataloader = single_dataloader()
+        self.num_data = len(self.dataloader)
+        self.model = heuristic()
```
For a more general analysis class, should take the test set as an input with the model.

TensorDuck commented on model.py:
```diff
@@ -0,0 +1,53 @@
+        return 1/(1+torch.exp(-inputs))
+
+
+class LSTM_baseline(nn.Module):
```
To make this module more modular, you should have all models follow a uniform format. I.E., they should all have a `fit()` and `inference()` method implemented. This way, when you want to try a different model, you just swap it out in the code and it will "just work." Otherwise, you'll need a custom notebook for every model, and that just doubled the amount of work you need for testing different models.

TensorDuck commented:
@henryhao1991 
Things you did well: 
* Your code looks efficient, making obvious use of previously imported functions rather than writing your own for the most part.
* You clearly know all the tricks for implementing clever functions/classes. This was something I think many fellows struggle with initially. Glad to see you likely wouldn't be.

So as mentioned, for a coder at your level, I'm primarily looking for comments and ways to improve the overall readability of your code. 
Your Loader method, for example, looks to do a lot of different things with several different classes. It just feels like this module can be made in a much simpler way (and definitely clearer) way. Comments would help, but ultimately I think the number of classes feels like too many for just trying to load a series of files into Tensors. 

Fix what you have time for, and then merge into master when ready. If you're not 100% sure on something, ignore it for now and make a note to ask me about it at our next check in. Keep going forward!

henryhao1991 committed:
Added LSTM training results

henryhao1991 committed:
Modified dataloader.py. Comments added for dataloader.py

henryhao1991 committed:
Code structure changed. TBD:LSTM training class;wrap up hero2vec files

henryhao1991 committed:
LSTM baseline training class added.

henryhao1991 committed:
Documentations and comments added for the codes.

henryhao1991 commented:
I modified the code structures and added comments for the code. I will just merge this branch now and open a new branch, where I will put the hero2vec code and results. If there is any bug with the existing code, they will be fixed in the new branch.
